lib/liquid_voting/delegations.ex

```diff
@@ -142,6 +142,48 @@ defmodule LiquidVoting.Delegations do
     |> Repo.insert()
   end
 
+  @doc """
+  Upserts a delegation (updates or inserts).
+
+  Updates existing global delegation for a specific delegator
+  if a new global delegation for the same delegator is passed in.
+  Updates existing proposal-specific delegation for a specific delegator
+  if a delegation for the same proposal and delegator is passed in.
+  Creates a new delegation if neither aforementioned condition is true.
+
+  ## Examples
+
+      iex> upsert_delegation(%{field: value})
+      {:ok, %Delegation{}}
+
+      iex> upsert_delegation(%{field: bad_value})
+      {:error, %Ecto.Changeset{}}
+  """
+  def upsert_delegation(attrs \\ %{}) do
+    %{delegator_id: delegator_id} = attrs
+    proposal_url = get_or_set_proposal(attrs)
+
+    conflicting_delegation =
+      Delegation
+      |> where([d], d.delegator_id == ^delegator_id)
+      |> where_proposal(proposal_url)
+      |> Repo.one()
+
+    upsert(conflicting_delegation, attrs)
+  end
+
+  defp get_or_set_proposal(%{proposal_url: proposal_url}), do: proposal_url
+  defp get_or_set_proposal(_attrs), do: nil
+
+  defp where_proposal(query, _proposal_url = nil),
+    do: query |> where([d], is_nil(d.proposal_url))
+
+  defp where_proposal(query, proposal_url),
+    do: query |> where([d], d.proposal_url == ^proposal_url)
+
+  defp upsert(_delegation = nil, attrs), do: create_delegation(attrs)
+  defp upsert(delegation, attrs), do: update_delegation(delegation, attrs)
```

Comment:
I wonder if we could replace these with `Repo.insert_or_update/2`?
https://hexdocs.pm/ecto/Ecto.Repo.html#c:insert_or_update/2?

Reply:
Interesting. Didn't know that is a thing. Will try.

Reply:
This will take me a while, as the example uses:   
```
result =
  case MyRepo.get(Post, id) do
    nil  -> %Post{id: id} # Post not found, we build one
    post -> post          # Post exists, let's use it
  end
  |> Post.changeset(changes)
  |> MyRepo.insert_or_update
```
So, I will need to find a way to wrap my query in a case statement, etc.

Reply:
I think you can pipe into it:

```elixir
Delegation
      |> where([d], d.delegator_id == ^delegator_id)
      |> where_proposal(proposal_url)
      |> Repo.one()
      |> case do
          nil -> ...
          delegation -> ...
      end
      |> ....

Though in this case, it might be preferable to keep the case logic in the name functions (again personal preference based on how easy it is to understand the code).
```

Reply:
OK, but where do I get the `changes` var from in the example?

Reply:
Also, I have different queries, based on whether proposal_url is nil.

Reply:
Ah, wait, that would be `attrs` right?

Reply:
Seems to work!! :)

Should I add the...
```
case result do
  {:ok, struct}       -> # Inserted or updated with success
  {:error, changeset} -> # Something went wrong
```
?

Reply:
I don't think there's a need to match on the result as we don't have any custom error handling for now (other than at the GraphQL layer)